fix(widget): update tokens based on widget params

alfetopito opened the pull request:
# Summary

Part of https://github.com/cowprotocol/cowswap/issues/4321

Before this change, tokens update from the widget were applied only once.
That's because the widget params are cached, but the app url (tokens are stored in the url) is updated independently.
Thus, when trying to update the app url for a second time, the update was ignored as it was identical to previous url coming from the widget.

# To Test

I don't know of an easy way from this PR, as the behaviour is not reproducible in the configurator.

I tested using a local version of https://examples.widget.cow.fi/ pointing to my local cowswap version.
Steps in this case are as follows:

1. Using the code from https://github.com/cowprotocol/widget-examples/pull/17, set the `baseUrl` widget param to this PR's CoW Swap version.
2. Start the webpack widget example from it
3. Connect the wallet to mainnet
4. Change tokens using the dropdowns
5. Click on clear tokens
* Should select weth as sell and buy should be empty
6. Select new tokens
7. Repeat `5`
* Same result

elena-zh reviewed (approved):
Not sure how to test this from UI, but I removed defaults in the tokens params
![image](https://github.com/user-attachments/assets/8ffd35ac-bc88-40f9-88a8-e82c31ad91ca)
, placed some trades and changed  networks: default state was applied to the form

alfetopito commented:
> Not sure how to test this from UI, but I removed defaults in the tokens params ![image](https://private-user-images.githubusercontent.com/70885163/360903804-8ffd35ac-bc88-40f9-88a8-e82c31ad91ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MTEyOTYsIm5iZiI6MTcyNDQxMDk5NiwicGF0aCI6Ii83MDg4NTE2My8zNjA5MDM4MDQtOGZmZDM1YWMtYmM4OC00MGY5LTg4YTgtZTgyYzMxYWQ5MWNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDExMDMxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjYTZlN2VkNWZmMjA1N2I3OGU3Y2FhN2IxMzlkNzkzNWUyN2NmMTBjZjg5ODU2NTkxZWRiNDBkZWU2Nzg2MjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bcevs29vNjOJG8n7WHuWo9dXFC_KKE7ovXgKlwSGIUw) , placed some trades and changed networks: default state was applied to the form

That works currently on widget.cow.fi, which doesn't have this fix.
I could observe it only in the described steps running the example locally :/

alfetopito commented on apps/cowswap-frontend/src/modules/trade/hooks/setupTradeState/useSetupTradeState.ts:
```diff
@@ -124,7 +125,9 @@ export function useSetupTradeState(): void {
     const tokensAreEmpty = !inputCurrencyId && !outputCurrencyId
 
     const sameTokens =
-      (inputCurrencyId || outputCurrencyId) && inputCurrencyId?.toLowerCase() === outputCurrencyId?.toLowerCase()
+      inputCurrencyId !== EMPTY_TOKEN_ID &&
+      (inputCurrencyId || outputCurrencyId) &&
+      inputCurrencyId?.toLowerCase() === outputCurrencyId?.toLowerCase()
```
This is the main fix. Otherwise the app never clears out the inner state.
Widget consumers (Den) want to be able to clear out both sell & buy tokens, which sets their values to `_`.

alfetopito commented on apps/cowswap-frontend/src/modules/injectedWidget/updaters/InjectedWidgetUpdater.tsx:
```diff
@@ -87,13 +87,24 @@ export function InjectedWidgetUpdater() {
   useEffect(() => {
     // Stop listening of message outside of React
     window.removeEventListener('message', cacheMessages)
 
     // Start listening for messages inside of React
     const updateParamsListener = listenToMessageFromWindow(window, WidgetMethodsListen.UPDATE_PARAMS, (data) => {
-      if (prevData.current && deepEqual(prevData.current, data)) return
+      if (
+        // If the data is the same as the previous data
+        prevData.current && deepEqual(prevData.current, data) &&
+        // And the pathname is the same as the current widget pathname, do nothing
+        // This is needed since the app updates the pathname independently of the widget params
+        pathNameRef.current === data.urlParams.pathname) {
+        return
+      }
```
This is the other fix, that allows the update to be triggered even when the app params don't change.
That's the case when the token selection is done in the app rather than via the widget.
